chart: fix service account name

kd7lxl opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->
/kind bug

#### What this PR does / why we need it:

The scheduler name, and subsequently the scheduler service account name can be configured by a value, but the deployment still referenced a hardcoded service account name. This resulted in the deployment failing whenever the scheduler name was overridden.

https://github.com/kubernetes-sigs/scheduler-plugins/blob/ac60529f9d8f8d444c952352876571d5b99d49d5/manifests/install/charts/as-a-second-scheduler/templates/serviceaccount.yaml#L4

https://github.com/kubernetes-sigs/scheduler-plugins/blob/ac60529f9d8f8d444c952352876571d5b99d49d5/manifests/install/charts/as-a-second-scheduler/templates/deployment.yaml#L41

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

zwpaper commented:
/area helm
/assign
/ok-to-test

Huang-Wei commented:
Strictly speaking, it's not a bug :)

/remove-kind bug
/kind enhancement

Could you also update the release-note section to:

```
Helm: tie controller/scheduler's service account name to their variable name.
```

kd7lxl commented on manifests/install/charts/as-a-second-scheduler/templates/deployment.yaml:
```diff
@@ -38,7 +38,7 @@ spec:
         component: scheduler
     spec:
-      serviceAccountName: scheduler-plugins-scheduler
+      serviceAccountName: {{ .Values.scheduler.name }}
```
The service account's name is currently working, so expanding the scope to parameterize it would make this an enhancement. 😄

kd7lxl replied:
Updated.

zwpaper commented:
hi @kd7lxl, I have found some time to look into the PR, and I found that you use the `scheduler.name` to ack as the serviceAccountName for the scheduler, but it seems that having a rbac related configuration would be better.

could you please add a `scheduler.serviceAccount.name` to specify the serviceAccountName, like

```yaml
scheduler:
  serviceAccount:
    name:
```

and later, we can raise another PR, or keep it in the one that would also be great, to implement the RBAC related in our chart.

ref: https://helm.sh/docs/chart_best_practices/rbac/

WDYT @Huang-Wei

zwpaper reviewed (changes requested):
/kind feature

kd7lxl commented:
> hi @kd7lxl, I have found some time to look into the PR, and I found that you use the `scheduler.name` to ack as the serviceAccountName for the scheduler, but it seems that having a rbac related configuration would be better.
> 
> could you please add a `scheduler.serviceAccount.name` to specify the serviceAccountName, like
> 
> ```yaml
> scheduler:
>   serviceAccount:
>     name:
> ```
> 
> and later, we can raise another PR, or keep it in the one that would also be great, to implement the RBAC related in our chart.
> 
> ref: https://helm.sh/docs/chart_best_practices/rbac/
> 
> WDYT @Huang-Wei

Sorry, I'm not sure what you're requesting. The chart already has RBAC, and already uses the parameterized name. That's why this was marked /kind bug. The names need to match everywhere for it to work.

https://github.com/kubernetes-sigs/scheduler-plugins/blob/ac60529f9d8f8d444c952352876571d5b99d49d5/manifests/install/charts/as-a-second-scheduler/templates/rbac.yaml#L87

I would prefer to keep any enhancements, like the ability to customize the service account name separate from the deployment name, in a separate PR so that this one is limited to the bug fix.

zwpaper commented:
Oops, sorry for the mistake...

what I say is that we should expose the rbac related in `values.yaml` like it mentioned here https://helm.sh/docs/chart_best_practices/rbac/.

it's ok for this PR to go with the enhancement with this `serviceAccountName`

zwpaper commented on .github/workflows/chart.yaml:
```diff
@@ -0,0 +1,48 @@
+      - name: Run chart-testing (list-changed)
+        id: list-changed
+        run: |
+          changed=$(ct list-changed --target-branch ${{ github.event.repository.default_branch }})
```
we may need to specify the `--chart-dirs`

kd7lxl replied:
I set the working directory to achieve this. Need to run it to know if it works this way. Not sure how to trigger a test run in this project.

zwpaper commented on manifests/install/charts/as-a-second-scheduler/ci/default-values.yaml:
why is there 2 empty default yaml?

kd7lxl replied:
Oops, fixed.

kd7lxl commented:
I added a chart test with a test case for the name override feature that is currently broken in the default branch. If the test passes, we will know this PR fixes the problem. The test was implemented in Github Actions, and I'm not sure how to run it in this project. It appears the project does not yet utilize Github Actions, but from past experience, that's the easiest tool for implementing chart testing.

Huang-Wei commented:
> Can someone help?

Now it's running. Let's wait and see.

Huang-Wei commented on manifests/install/charts/as-a-second-scheduler/ci/default-values.yaml:
this file is empty, is it intended?

kd7lxl replied:
Yes, it tells `ct` to run a test case with the default values (no overrides).

Huang-Wei commented on .github/workflows/chart.yaml:
```diff
@@ -0,0 +1,48 @@
+name: Lint and Test Chart
```
nit: maybe adding "helm" to it? like "Helm Chart Lint and Test"

Huang-Wei commented:
Thanks @kd7lxl for the Helm testing bits. It runs successfully.

To keep a linear git history, would you mind squashing the commits into two: one is core changes, the other for helm CI?

kd7lxl commented:
> > Can someone help?
> 
> Now it's running. Let's wait and see.

Hmm, it looks like it [ran](https://github.com/kubernetes-sigs/scheduler-plugins/actions/runs/4631296377/jobs/8195809018?pr=570), but skipped the test cases. Need to figure that out.

zwpaper commented:
> Hmm, it looks like it [ran](https://github.com/kubernetes-sigs/scheduler-plugins/actions/runs/4631296377/jobs/8195809018?pr=570), but skipped the test cases. Need to figure that out.

maybe you need to define the output, and `need` dependency first?
https://docs.github.com/en/actions/using-jobs/defining-outputs-for-jobs

Huang-Wei commented:
@kd7lxl if it's too much work, I'm fine with only focus on parameterize the service account in this PR, and follow up the CI bits in another PR. But it's up to you :)

kd7lxl commented:
> To keep a linear git history, would you mind squashing the commits into two: one is core changes, the other for helm CI?

> @kd7lxl if it's too much work, I'm fine with only focus on parameterize the service account in this PR, and follow up the CI bits in another PR. But it's up to you :)

Sorry, was away from the computer for a few days. Squashed and rebased now to just the service account name fix. I'll open another PR to add chart tests.

Huang-Wei commented:
/retest

Huang-Wei commented:
/lgtm
/approve
